File: problems/0035 - Search Insert Position/Kotlin/Solution.kt
class Solution {
    fun searchInsert(nums: IntArray, target: Int): Int {
        var left = 0
        var right = nums.size - 1
        var mid = (left + right) / 2

        while (left <= right) {
            when {
                nums[mid] == target -> return mid
                nums[mid] > target -> right = mid - 1
                else -> left = mid + 1
            }
            mid = (left + right) / 2
        }

        return if(nums[mid] > target) mid else mid + 1
    }
}

File: problems/0035 - Search Insert Position/Go/Solution.go
// Runtime: 4 ms
// Memory Usage: 3 MB

func searchInsert(nums []int, target int) int {
	start := 0
	end := len(nums) - 1
	var mid int
	for start <= end {
		mid = (start + end + 1) / 2
		if nums[mid] == target {
			return mid
		}
		if nums[mid] < target {
			start = mid + 1
		} else {
			end = mid - 1
		}
	}
	return start
}